stan/math/rev/fun/beta.hpp

```diff
@@ -34,197 +34,42 @@ namespace math {
  * @param b var Argument
  * @return Result of beta function
  */
-inline var beta(const var& a, const var& b) {
-  double digamma_ab = digamma(a.val() + b.val());
-  double digamma_a = digamma(a.val()) - digamma_ab;
-  double digamma_b = digamma(b.val()) - digamma_ab;
-  return make_callback_var(beta(a.val(), b.val()),
-                           [a, b, digamma_a, digamma_b](auto& vi) mutable {
-                             const double adj_val = vi.adj() * vi.val();
-                             a.adj() += adj_val * digamma_a;
-                             b.adj() += adj_val * digamma_b;
-                           });
-}
-
-/**
- * Returns the beta function and gradient for first var input.
- *
-   \f[
-     \mathrm{beta}(a,b) = \left(B\left(a,b\right)\right)
-   \f]
+template <typename T1, typename T2,
+          require_all_not_std_vector_t<T1, T2>* = nullptr,
+          require_return_type_t<is_var, T1, T2>* = nullptr>
+inline auto beta(const T1& a, const T2& b) {
+  arena_t<ref_type_t<T1>> arena_a = a;
+  arena_t<ref_type_t<T2>> arena_b = b;
 
-   \f[
-    \frac{\partial }{\partial a} = \left(\psi^{\left(0\right)}\left(a\right)
-                                      - \psi^{\left(0\right)}
-                                      \left(a + b\right)\right)
-                                    * \mathrm{beta}(a,b)
-   \f]
- *
- * @param a var Argument
- * @param b double Argument
- * @return Result of beta function
- */
-inline var beta(const var& a, double b) {
-  auto digamma_ab = digamma(a.val()) - digamma(a.val() + b);
-  return make_callback_var(beta(a.val(), b), [a, digamma_ab](auto& vi) mutable {
-    a.adj() += vi.adj() * digamma_ab * vi.val();
-  });
-}
+  const auto& beta_val = beta(value_of(arena_a), value_of(arena_b));
+  using return_type_t = return_var_matrix_t<decltype(beta_val), T1, T2>;
+  arena_t<return_type_t> res(beta_val);
 
-/**
- * Returns the beta function and gradient for second var input.
- *
-   \f[
-     \mathrm{beta}(a,b) = \left(B\left(a,b\right)\right)
-   \f]
+  reverse_pass_callback([arena_a, arena_b, res]() mutable {
+    auto&& a_array = as_array_or_scalar(arena_a);
+    auto&& b_array = as_array_or_scalar(arena_b);
+    const auto& res_array = as_array_or_scalar(res);
+    const auto& digamma_ab = digamma(value_of(a_array) + value_of(b_array));
+    const auto& adj_val = res_array.adj() * res_array.val();
```

Comment:
This produces a new expression objet so we don't need a `const&` of it
```suggestion
    auto adj_val = res_array.adj() * res_array.val();
```

Anywhere that we will produce a new Eigen expression it should just be an object.

Reply:
Same for several places here